server/src/main/java/io/crate/execution/ddl/tables/DropColumnTask.java

```diff
@@ -0,0 +1,198 @@
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.IndexTemplateMetadata;
+import org.elasticsearch.cluster.metadata.MappingMetadata;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.common.settings.IndexScopedSettings;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.index.mapper.DocumentMapper;
+import org.elasticsearch.index.mapper.MapperService;
+
+import com.carrotsearch.hppc.IntArrayList;
+
+import io.crate.analyze.AlterTableDropColumnAnalyzer;
+import io.crate.common.CheckedFunction;
+import io.crate.common.collections.Lists2;
+import io.crate.common.collections.Maps;
+import io.crate.exceptions.ColumnUnknownException;
+import io.crate.metadata.NodeContext;
+import io.crate.metadata.PartitionName;
+import io.crate.metadata.Reference;
+import io.crate.metadata.cluster.DDLClusterStateHelpers;
+import io.crate.metadata.cluster.DDLClusterStateTaskExecutor;
+import io.crate.metadata.doc.DocTableInfo;
+import io.crate.metadata.doc.DocTableInfoFactory;
+
+public final class DropColumnTask extends DDLClusterStateTaskExecutor<DropColumnRequest> {
+
+    private static final IntArrayList EMPTY_PK_LIST = new IntArrayList();
+
+    private final NodeContext nodeContext;
+    private final CheckedFunction<IndexMetadata, MapperService, IOException> createMapperService;
+
+    public DropColumnTask(NodeContext nodeContext, CheckedFunction<IndexMetadata, MapperService, IOException> createMapperService) {
+        this.nodeContext = nodeContext;
+        this.createMapperService = createMapperService;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public ClusterState execute(ClusterState currentState, DropColumnRequest request) throws Exception {
+        DocTableInfoFactory docTableInfoFactory = new DocTableInfoFactory(nodeContext);
+        DocTableInfo currentTable = docTableInfoFactory.create(request.relationName(), currentState);
+
+        List<DropColumn> normalizedColumns = AlterTableDropColumnAnalyzer.validateDynamic(
+            currentTable, normalizeColumns(request, currentTable));
+
+        if (normalizedColumns.isEmpty()) {
+            return currentState;
+        }
+
+        var refsToDrop = Lists2.map(normalizedColumns,
+                                    dc -> {
+                                        dc.ref().setDropped();
+                                        return dc.ref();
+                                    });
+
+        Metadata.Builder metadataBuilder = Metadata.builder(currentState.metadata());
+        Map<String, Object> mapping = createMapping(
+            MappingUtil.AllocPosition.forTable(currentTable),
```

Comment:
I think we need to propagate top-level drop flag to sub-column (AFAIK we agreed to drop the whole sub-tree when deleting an object column). Could you add a test that dropping column o also makes o[a] and o[a][b] invisible/dropped?